MetaDraw change color of variant containing ions

cameyer6 opened the pull request.

rmillikin commented on EngineLayer/PsmTsv/PsmFromTsv.cs:
```diff
@@ -201,5 +209,43 @@ private static List<MatchedFragmentIon> ReadFragmentIonsFromString(string matche
+                        Match ionMatch = IonParser.Match(ion.Annotation);
+                        if (ionMatch.Success && // finds match, variant is within peptide, and ion contains variant (differs if ion is abc or xyz)
+                            (variantResidueEnd >= peptideStart && variantResidueStart <= peptideEnd) &&
+                            ((ion.NeutralTheoreticalProduct.ProductType <= ProductType.c &&
```
this `<= ProductType.c` stuff seems weird

lonelu replied:
I guess (ProductType <= ProductType.c) is same as (ProductType ==ProductType.a || ProductType ==ProductType.b || ProductType ==ProductType.c). Please make sure it works like what it means.

cameyer6 replied:
It now checks if the ProductType is in a list of a, b, and c product types or x, y, and z product types.

rmillikin commented on GUI/MetaDraw.xaml.cs:
```diff
@@ -476,6 +476,7 @@ private void DrawAnnotatedBaseSequence(PsmFromTsv psm)
             {
                 int residue = ion.NeutralTheoreticalProduct.TerminusFragment.AminoAcidPosition;
                 string annotation = ion.NeutralTheoreticalProduct.ProductType + "" + ion.NeutralTheoreticalProduct.TerminusFragment.FragmentNumber;
+                Color color = psm.VariantCrossingIons.ContainsKey(ion) && psm.VariantCrossingIons[ion] ? Colors.Green : productTypeToColor[ion.NeutralTheoreticalProduct.ProductType];
```
`&& psm.VariantCrossingIons[ion]` doesn't do anything

rmmiller22 replied:
It does check to see if the entry with the key matched ion crosses the variant site. The dictionary is poorly named because it contains all of the matched ions, those that do and those that do not cross the variant site.

cameyer6 replied:
VariantCrossingIons a List now, so that expression is now `psm.VariantCrossingIons.Contains(ion)`.

rmillikin commented on GUI/MetaDraw/PsmAnnotationViewModel.cs:
```diff
@@ -293,7 +293,11 @@ public void DrawPeptideSpectralMatchPdf(MsDataScan msDataScan, PsmFromTsv psm, s
             OxyColor ionColor;
 
-            if (productTypeDrawColors.ContainsKey(matchedIon.NeutralTheoreticalProduct.ProductType))
+            if(psmToDraw.VariantCrossingIons.ContainsKey(matchedIon) && psmToDraw.VariantCrossingIons[matchedIon])
```
` && psmToDraw.VariantCrossingIons[matchedIon]` doesn't do anything

rmmiller22 replied:
same response as the comment above

rmmiller22 commented on Test/MetaDrawTest.cs:
```diff
@@ -54,5 +55,40 @@ public static void TestMetaDrawReadCrossPsmFile()
+            List<string> warnings = new List<string>();
+            List<PsmFromTsv> psms;
+
+            psms = PsmTsvReader.ReadTsv(myFile, out warnings);  //test will fail if order of psms isn't from top to bottom row
```
Check that you are parsing everything in correctly now that you have added the identified variant field. Add assert statements to ensure that these are as expected.

cameyer6 replied:
I added checks for Intersecting Sequence Variations, Identified Sequence Variations, and Splice Sites.

rmmiller22 commented on Test/MetaDrawTest.cs:
```diff
@@ -54,5 +55,40 @@ public static void TestMetaDrawReadCrossPsmFile()
+            expected.Add(new Dictionary<string, bool>() { { "b2", false }, { "b4", false }, { "y1", false }, { "y14", false } });    // variant after peptide (6532)
+
+
+            Assert.AreEqual(0, psms[0].VariantCrossingIons.Count, 
```
Don't just check for one psm. Ensure that the count is as expected for all the psms

cameyer6 replied:
The whole unit test works differently now that VariantCrossingIons is a List, but it checks that there aren't any extra ions in the list.

rmmiller22 commented on EngineLayer/PsmTsv/PsmFromTsv.cs:
```diff
@@ -38,13 +40,16 @@ public class PsmFromTsv
         public string DecoyContamTarget { get; }
         public double? QValueNotch { get; }
 
+        public Dictionary<MatchedFragmentIon, bool> VariantCrossingIons { get; }
```
The dictionary should be renamed it is confusing because it seems that this dictionary would only contains fragment ions that cross the variant, which is not the case.

lonelu replied:
Is that possible to add a property to MatchedFragmentIon, like "bool ionCrossingVariant {get; set;}" ? So no extra dictionary is needed here.

rmmiller22 replied:
That was the original plan but Rob wanted to avoid making changes to mzlib, which that required.

trishorts commented:
I think that we should add these three psm fields to the Properties window of the metadraw output (Intersecting Sequence   Variations,  Identified Sequence Variations ,  Splice Sites).

![image](https://user-images.githubusercontent.com/16841846/60524370-085dfd80-9cb2-11e9-906f-bf35ccd09f46.png)

trishorts commented:
Also, this is not specific to variants, but I have a hard time seeing the yellow dot at the front of this annotated sequence. maybe a slightly darker color like orange? 
![image](https://user-images.githubusercontent.com/16841846/60525735-8cb18000-9cb4-11e9-8f85-becc6853ea99.png)

cameyer6 commented:
> I think that we should add these three psm fields to the Properties window of the metadraw output (Intersecting Sequence Variations, Identified Sequence Variations , Splice Sites).
> 
> ![image](https://user-images.githubusercontent.com/16841846/60524370-085dfd80-9cb2-11e9-906f-bf35ccd09f46.png)

I added those properties and changed the modification color on the annotated base sequence to orange.

trishorts commented:
seeing some strange things related to variants in metadraw. just loaded a tsv with only two variant psms from a raw file but I see five scans listed. first three are incorrect but last two are correct.
![image](https://user-images.githubusercontent.com/16841846/61299230-4e808a00-a7a5-11e9-9dae-30cd711d82f4.png)

in the properties window, variantcrossingions is not displaying the ions, but instead an error 
![image](https://user-images.githubusercontent.com/16841846/61299342-7ff95580-a7a5-11e9-8db2-fc7f5726c201.png)
